File: gameOfLife.py
lifeFile = open('life.txt','r')

# variable declarations
i = 0 
replacedFile = ""
addSpace = "" 

# get line count
for line in lifeFile:


	# for first line with rows & cols count
	if (i == 0) :
		rowCols = line.split()
		rows = int(rowCols[0])
		cols = int(rowCols[1])

	else :
		
		# check in range
		if ( i < rows ):

			replacedLine1 = line.replace(" ","-")
			replacedLine2 = "-"+replacedLine1.replace("\n","-") #adding padding

			lineLength = len(replacedLine2)

			if(lineLength < cols):
				addSpace = '-'+('-'*(cols-lineLength))
			
			replacedLine = replacedLine2 + addSpace
			replacedFile = replacedFile + replacedLine + "\n"

	# increment line count
	i = i + 1

"""
	add extra lines with padding 
"""
addExtraLine = ""
addExtraLine = "-"*(cols+1) + "\n" 

""" 
	this is the final file with dashes and stars 
	adding extra two lines in the file

"""
replacedFile = addExtraLine  + replacedFile + addExtraLine + addExtraLine


"""
spliting the lines in the replacedFile to a list
"""
splitLines = replacedFile.splitlines()

"""
create a list of 
alivePositions : list of all the stars(*)
deadPositions  : list of all the dashes(-)
aliveNeighbors : list of all the neighbors that are alive
"""
alivePositions = [[] for j in range (len(splitLines))]
deadPositions  = [[] for j in range (len(splitLines))]

for i in range(len(splitLines)):

	for cells in range(len(splitLines[i])):

		if (splitLines[i][cells] == "*"):
			alivePositions[i].append(cells)
			deadPositions[i].append('1')

		if (splitLines[i][cells] == "-"):
			deadPositions[i].append('0')


"""
	check the neighbors for cell (x,y)

	[x-1,y-1]		[x-1,y] 		[x-1,y+1] 
	 [x,y-1] 		 [x,y] 			 [x,y+1]
	[x+1,y-1] 		[x+1,y] 		[x+1,y+1]

"""
def getNeighbors(row,cell):

	# get x and y co-ordinate
	x = row
	y = cell

	#create an empty list(neighbors) of 8 rows
	neighbors = [[] for i in range (len(splitLines))]

	neighbors[x-1].append(y-1)
	neighbors[x-1].append(y)
	neighbors[x-1].append(y+1)

	neighbors[x].append(y-1)
	neighbors[x].append(y+1)

	neighbors[x+1].append(y-1)
	neighbors[x+1].append(y)
	neighbors[x+1].append(y+1)
	
	return neighbors

"""
get the list of neighbors that are alive
"""
def getAliveNeighbors(neighborsList,alivePositionsList,rowNum):
	
	aliveNeighbors = [[] for j in range (len(splitLines))]

	#for same row
	for neighborCell in neighborsList[rowNum]:
		for aliveCell in alivePositionsList[rowNum]:
			if neighborCell == aliveCell :
				if aliveCell not in aliveNeighbors[rowNum]:
					aliveNeighbors[rowNum].append(aliveCell)

	#for prev row
	for neighborCell in neighborsList[rowNum-1]:
		for aliveCell in alivePositionsList[rowNum-1]:
			if neighborCell == aliveCell :
				if aliveCell not in aliveNeighbors[rowNum-1]:
					aliveNeighbors[rowNum-1].append(aliveCell)

	#for next row
	for neighborCell in neighborsList[rowNum+1]:
		for aliveCell in alivePositionsList[rowNum+1]:
			if neighborCell == aliveCell :
				if aliveCell not in aliveNeighbors[rowNum+1]:
					aliveNeighbors[rowNum+1].append(aliveCell)

	return aliveNeighbors

"""
get list of the neighbors of the dead cells

[x-1,y-1]		[x-1,y] 		[x-1,y+1] 
 [x,y-1] 		 [x,y] 			 [x,y+1]
[x+1,y-1] 		[x+1,y] 		[x+1,y+1]

"""


"""
A living cell with two or three neighboring living cells survives into the next generation.
A living cell with fewer than two or more than three living neighbors will die.
"""
def isAlive(aliveNeighbors):
	count = 0 
	for row in aliveNeighbors:
		for cell in row :
			if(cell):
				count += 1

	if(count == 2 or count == 3):
		return 1
	else :
		return 0


"""
for every cell in the alivePositions list check aliveNeighbors
"""
aliveLineNum = 0
nextGenAliveList = [[] for j in range (len(splitLines))]

for row in alivePositions :
	
	for cell in row :
		
		# generate list of neighbors that will be for a cell
		neighborsList = getNeighbors(aliveLineNum,cell)
		
		"""
		for the particular cell generate list of Neighbors that are alive
		by finding intersection between the neighbors it has 
		and the list of neighbors that are alive
		"""
		aliveNeighbors = getAliveNeighbors(neighborsList,alivePositions,aliveLineNum)

		"""
		for each cell check if it will be alive
		a cell is alive if 2 or 3 neighbors are alive
		"""
		isAliveStatus = isAlive(aliveNeighbors)
		
		if(isAliveStatus):
			nextGenAliveList[aliveLineNum].append(cell)

	aliveLineNum = aliveLineNum+1

"""
	for every cell in deadPositions list get neighbors
"""
def isDead(x,y,listName):
	if (listName[int(x)][int(y)] == '0') :
		return True
	else :
		return False


def getNeighborsListXY(x,y):
	listXY = []
	neighborsList = getNeighbors(int(x),int(y))
	
	rowNum = 0
	for row in neighborsList :
		for cell in row :
			if(cell):
				listXY.append((rowNum,cell))

		rowNum += 1

	return listXY

modRowNum = 0
print deadPositions
for row in deadPositions :
	
	if(modRowNum == 0):
		pass 
	else:
		for col in row:

			if (isDead(modRowNum,col,deadPositions)) :
				try :
					neighbors = getNeighborsListXY(modRowNum,col)
					
				except IndexError:
					pass

	modRowNum += 1

print "----"
print alivePositions

File: gameOfLife.java
import java.io.File ;
import java.util.* ;

public class gameOfLife
{
    public static int rowsCount = 0;
    public static int colsCount = 0;

    public static void main(String[] args)
        throws java.io.FileNotFoundException
    {  
        int iterations;

        if(args.length == 0)
           iterations = 10;
        else
           iterations = Integer.parseInt(args[0]);
       
        ArrayList<ArrayList<Boolean>> gameState = readFile("life.txt");
        ArrayList<int[]> alivePositions = alivePositions(gameState);
        ArrayList<int[]> deadPositions  = deadPositions(gameState);
        ArrayList<int[]> willSustain  = willSustain(alivePositions,gameState);
        ArrayList<int[]> willFlourish = willFlourish(deadPositions,gameState);
        ArrayList<ArrayList<Boolean>> emptyGrid = emptyGrid();
        ArrayList<ArrayList<Boolean>> nextGen = setNextGen(emptyGrid,willSustain);
        ArrayList<ArrayList<Boolean>> newNextGen = setNextGen(nextGen,willFlourish);  
        printStates(newNextGen);

        for(int i = 0 ; i < iterations-1 ; i++ )
        {
            alivePositions = alivePositions(newNextGen);
            deadPositions  = deadPositions(newNextGen);
            willSustain  = willSustain(alivePositions,newNextGen);


            willFlourish = willFlourish(deadPositions,newNextGen);
            emptyGrid = emptyGrid();
            nextGen = setNextGen(emptyGrid,willSustain);

            newNextGen = setNextGen(nextGen,willFlourish);  
            printStates(newNextGen);
        }

    }


    /*--
    Reading the input file : life.txt
    @param : fileName
    @return : gameState which consists of the array list of positions
    --*/
    public static ArrayList<ArrayList<Boolean>> readFile(String fileName)
        throws java.io.FileNotFoundException
    {

        ArrayList<String> lines = new ArrayList<String>();
        Scanner scan = new Scanner(new File(fileName));

        while(scan.hasNextLine())
            lines.add(scan.nextLine());
        
        scan.close();
        
        String[] lineArray = new String[lines.size()];

        // convert ArrayList to Array of defined size
        lines.toArray(lineArray);

        // We will use pass this gameState throughout the program.
        ArrayList<ArrayList<Boolean>> gameState = null;
        

        boolean notFirstTime = false;
        int countOfLines = 0;

        for(String line:lines)
        {
            //check if its the First Line
            if(!notFirstTime){

                String[] rowColsCount = line.split(" ");
                rowsCount = Integer.parseInt(rowColsCount[0]);
                colsCount = Integer.parseInt(rowColsCount[1]);
                
                //We pass the rowsCount to the constructor of the arraylist
                gameState = new ArrayList<ArrayList<Boolean>>(rowsCount);
                notFirstTime = true;
            }
            else
            {
                String[] splitLineArray = line.split("");
                ArrayList<Boolean> tempArray = new ArrayList<Boolean>(colsCount);
                
                /*
                    for all the cells in the column set the sate as true and false 
                    if the size of the row is lesser than the column count then add false 
                    through the ArrayIndexOutOfBoundException
                 */
                for (int j= 1 ; j<=colsCount; j++)
                { 
                    try {

                        if(splitLineArray[j].equals("*")) {
                            tempArray.add(j-1,true);
                            
                        }
                        else
                            tempArray.add(j-1,false);
                    }
                    catch (ArrayIndexOutOfBoundsException e)
                    {
                        tempArray.add(j-1,false);
                    }
                }
                gameState.add(tempArray);
                countOfLines++;
            }

            

        }

        /* 
            if the size of the rows are less than the ones 
            then add extra empty rows to our game
        */
        while(gameState.size() < rowsCount )
        {
           ArrayList<Boolean> tempArray = new ArrayList<Boolean>(colsCount);
           for (int j= 0 ; j<colsCount; j++)
           {
                tempArray.add(j,false);
           }
           gameState.add(tempArray);
           
        } 

      return gameState;
    }

    
    /*  
    create the list of neighbors for points with x & y co-ordinates
        {x-1,y-1}       {x-1,y}         {x-1,y+1} 
         {x,y-1}         {x,y}           {x,y+1}
        {x+1,y-1}       {x+1,y}         {x+1,y+1}

    @param : the x and y co-ordinates of the cell
    @return : a list of neighbors
    */
   public static ArrayList<int[]> getNeighbors(int x, int y)
   {
        ArrayList<int[]> neighbors = new ArrayList<int[]>();

        int[] position1 = {x-1,y-1};
        int[] position2 = {x-1,y};
        int[] position3 = {x-1,y+1};

        int[] position4 = {x,y-1};
        int[] position5 = {x,y+1};

        int[] position6 = {x+1,y-1};
        int[] position7 = {x+1,y};
        int[] position8 = {x+1,y+1};

        if(x-1 >= 0 && y-1 >= 0)
            neighbors.add(position1); 

        if(x-1>=0)
            neighbors.add(position2);

        if(x-1>=0 && y+1 <colsCount)
            neighbors.add(position3);

        if(y-1>=0)
            neighbors.add(position4);

        if(y+1 < colsCount)
            neighbors.add(position5);

        if(x+1 < rowsCount && y-1>=0)
            neighbors.add(position6);
        
        if(x+1 < rowsCount)
            neighbors.add(position7);

        if(x+1 < rowsCount && y+1 < colsCount)
            neighbors.add(position8);

        return neighbors;
    }

    /*
        check if the cell is alive in the original gameState list
        if alive return Boolean

     */ 
    public static Boolean isAlive(int x,int y,ArrayList<ArrayList<Boolean>> gameState)
    {
        if( gameState.get(x).get(y) )
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    /*
        create a list of alivePositions in the gameState
        @param : The original gameState list
        @return : the list of positions alive in an arrayList tuple e.g. : [{1,2}{2,3}]
     */ 
    public static ArrayList<int[]> alivePositions(ArrayList<ArrayList<Boolean>> gameState)
    {
        int rowNumber = 0;
        ArrayList<int[]> alivePositions = new ArrayList<int[]>();

        for(ArrayList<Boolean> row : gameState)
        {
            
                      
            for(int i=0; i<row.size();i++)
            {
                if(row.get(i))
                {
                    int[] alivePoint = {rowNumber,i};
                    alivePositions.add(alivePoint);
                }
            }
            
            rowNumber++;
        }

        return alivePositions;
    }

    /*
        create a list of deadPositions in the gameState
        @param : The original gameState list
        @return : the list of positions dead in an arrayList tuple e.g. : [{1,2}{2,3}]
     */
    public static ArrayList<int[]> deadPositions(ArrayList<ArrayList<Boolean>> gameState)
    {
        int rowNumber = 0;
        ArrayList<int[]> deadPositions = new ArrayList<int[]>();

        for(ArrayList<Boolean> row : gameState)
        {

            for(int i=0; i<row.size();i++)
            {
                if(!row.get(i))
                {
                    int[] alivePoint = {rowNumber,i};
                    deadPositions.add(alivePoint);    
                }
            }

            
            rowNumber++;
        }

        return deadPositions;
    }

    /* 
    this will give a list of ALIVE cells that will stay alive for the next generation
    @param : the list of Alive Positions in the current game State list
    @return : the list of elements that will sustain for the next generation
    */ 
    public static ArrayList<int[]> willSustain(ArrayList<int[]> alivePositions,ArrayList<ArrayList<Boolean>> gameState)
    {
        ArrayList<int[]> willSustain = new ArrayList<int[]>() ;
        for( int i=0 ; i < alivePositions.size(); i++)
        {

                // for each cell with co-ordinates x and y
                int x = alivePositions.get(i)[0];
                int y = alivePositions.get(i)[1];

                // get neighbors
                ArrayList<int[]> neighbors = getNeighbors(x,y); 

                int countNeighbor = 0;
                for(int[] neighbor : neighbors)
                {   
                    if(isAlive(neighbor[0],neighbor[1],gameState))
                    {   
                        countNeighbor += 1;
                    }

                }

                if(countNeighbor == 2 || countNeighbor == 3)
                {
                    int[] addNeigbhor = {x,y};
                    willSustain.add(addNeigbhor);
                }
                
        }

        return willSustain;
    }

     /* 
    this will give a list of dead cells that will stay alive for the next generation
    @param : the list of dead Positions in the current game State list
    @return : the list of elements that will stay alive for the next generation
    */ 
    public static ArrayList<int[]> willFlourish(ArrayList<int[]> deadPositions,ArrayList<ArrayList<Boolean>> gameState)
    {
        ArrayList<int[]> willFlourish = new ArrayList<int[]>() ;
        for( int i=0 ; i < deadPositions.size(); i++)
        {
            // for each cell with co-ordinates x and y
            int x = deadPositions.get(i)[0];
            int y = deadPositions.get(i)[1];
            // System.out.println(x+" "+y);

            // get neighbors
            ArrayList<int[]> neighbors = getNeighbors(x,y); 

            //foreach neighbor if neighbor is alive increment counter 
            int countNeighbor = 0;
            for(int[] neighbor : neighbors)
            {   
                if(isAlive(neighbor[0],neighbor[1],gameState))
                {
                    countNeighbor += 1;
                }
            }

            if(countNeighbor == 3)
            {
                int[] addNeigbhor = {x,y};
                willFlourish.add(addNeigbhor);
            }
                
        }

        return willFlourish;
        
    }

    /*
        Create an empty grid of setting false positions
        @return : the empty grid in boolean value of cells to 'FALSE'

     */
    public static ArrayList<ArrayList<Boolean>> emptyGrid()
    {
        ArrayList<ArrayList<Boolean>> emptyGrid = new ArrayList<ArrayList<Boolean>>();
        for(int i = 0 ; i < rowsCount ; i++)
        {
            ArrayList<Boolean> rowArrayList = new ArrayList<Boolean>();
            for(int j=0; j < colsCount ; j++)
            {
                rowArrayList.add(j,false);
            }
            emptyGrid.add(rowArrayList);
        }

        return emptyGrid;
    }

    /*
        for the given grid set the variables to true or false for the next generation
        @param : the grid of positions with true-false set . the list of positions that you would want to set to true for the next generation
        @return : the grid that is formed from the list
     */
    public static ArrayList<ArrayList<Boolean>>  setNextGen(ArrayList<ArrayList<Boolean>> grid,ArrayList<int[]> givenArrayList)
    {   
        for(int[] cordo : givenArrayList)
        {
            ArrayList<Boolean> row = grid.get(cordo[0]);
            row.set(cordo[1],true);
        }

        return grid;
        
    }

    public static void printStates(ArrayList<ArrayList<Boolean>> gameState)
    {
        for(ArrayList<Boolean> row: gameState)
         {
             
            StringBuffer buffer = new StringBuffer();
            for(int i = 0 ; i<row.size();i++)
             {
                 if(row.get(i)==true)
                 {
                     buffer.append("*");
                 }
                 else
                 {
                    buffer.append("-");
                 }
            }

            System.out.println(buffer.toString());
            System.out.println();
            
         }

        System.out.println("===========================");
    }


}